lib/active_merchant/billing/gateways/visanet_peru.rb

```diff
@@ -197,15 +197,29 @@ def success_from(response)
       end
 
       def message_from(response, options, action)
-        if empty?(response['errorMessage']) || response['errorMessage'] == '[ ]'
-          action == 'refund' ? "#{response['data']['DSC_COD_ACCION']}, #{options[:error_message]}" : response['data']['DSC_COD_ACCION']
-        elsif action == 'refund'
-          message = "#{response['errorMessage']}, #{options[:error_message]}"
-          options[:error_message] = response['errorMessage']
-          message
-        else
-          response['errorMessage']
-        end
+        options[:error_message] = response['errorMessage'] if stash_error_message?(response, action)
+        message_from_messages(
+          response['errorMessage'],
+          action_code_description(response),
+          options[:error_message]
+        )
+      end
+
+      def message_from_messages(*args)
+        args.reject { |m| error_message_empty?(m) }.join(' | ')
```

Comment:
Does this change the error message format from commas to pipe? I'm not sure if this might cause a problem for customers if they are parsing on commas. Also, if it changes the order, we might think about reworking this.

Reply:
I don't think the joining was working in the first place: because the `cancelDeposit` case comes first, the original `elsif action == 'refund'` condition would not have been true such that the stashed message would have been returned to the user (it may have been stashed, but we would never have used it when the action was `refund`). This is what prompted my initial draft PR with the additional test to verify what the intended behavior was (properly distinguishing between a `refund` and `cancelDeposit` and how that affected the message construction.

If that is correct, my thinking was that the risk of end users relying on parsing the field is minimal, and the advantage of the pipe over the comma moving forward is that *if* sometimes the messages we're joining have punctuation already, I wanted to avoid something like "Error message., Error code explanation". I don't know whether that's a real concern, just noticed the punctuation in some of the canned responses.

Since there was only this one (probably not working) case where we were doing the joining of any kind, I don't think there is an order to preserve for the expanded case. In the specific case where we were joining only if the first `cancelDeposit` refund attempt failed, this does change the order. However, if people are parsing this string, then previously they would be getting the `response['errorMessage'] in practice, so putting that first seemed logical (even though it changes our original intended ordering in the special case).

I'm not opposed to changing any of this though, just laying out my thought process so far.

Reply:
Ok, that makes sense and if it wasn't working before there is a lower risk.